src/Datasource/RulesChecker.php

```diff
@@ -314,24 +314,37 @@ protected function _addError($rule, $name, $options)
 
         return function ($entity, $scope) use ($rule, $name, $options) {
             $pass = $rule($entity, $options + $scope);
-            if ($pass === true || empty($options['errorField'])) {
-                return $pass === true;
+            
+            if ($pass === true) {
+                return true;
             }
 
             $message = 'invalid';
+            $save = false;
             if (isset($options['message'])) {
                 $message = $options['message'];
+                $save = true;
             }
             if (is_string($pass)) {
                 $message = $pass;
+                $save = true;
             }
             if ($name) {
                 $message = [$name => $message];
             } else {
                 $message = [$message];
             }
-            $entity->errors($options['errorField'], $message);
-            return $pass === true;
+            
+            $errorField = '_rules';
+            if (isset($options['errorField'])) {
+                $errorField = $options['errorField'];
+                $save = true;
+            }
+            
+            if ($save) {
```

Comment:
Yes, but this would change the behaviour: Currently, when no message is specified, the default message "invalid" is used (when an errorField is set). When `$message` defaults to null, no message would be set if you only specify an errorField, but no message. The `$save` variable was the only way I found to keep this behaviour. (I didn't want to save "invalid" to the "_rules" field, if you neither specify errorField nor message.

But one could definitely simplify this (as the default message "invalid" is not documented, IIRC)